enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,235 @@
+
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+* User provide certificate and it expired after some time.
+  kube-api server will continue serving with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* User provided expired Certificate and Microshift service started/restarted 
+  Microshift will start with a warning in the logs, kube-apiserver will continue serve with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* Users might configure certs with the wrong names that doesnt match the certificate SAN,but openShift allows it so we will too.
+### Drawbacks
+* External certificate wont be automaticly renewed, so it requires manual Certificate rotation. 
+  similiar approach is taken by OpenShift.
+
+
+## Design Details
+N/A
+
+## Open Questions [optional]
+N/A
+
+## Test Plan
+add e2e test that will:
+- generate and sign certificates with custom ca.
+- change the default Microshift configuration to use the newly generated certs.
+- make sure system is functional using  generated external kubeconfig.
+- intentionally configure invalid value in `names` (ie: 127.0.0.1,localhost) , Microshift should ignore this configuration and skip this certificate.
+
+## Graduation Criteria
+### Dev Preview -> Tech Preview
+- Gather feedback from early adopters on possible issues.
+
+### Tech Preview -> GA
+- Extensive testing with various certificates variations.
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+N/A
+
+## Version Skew Strategy
+N/A
+
+## Operational Aspects of API Extensions
+N/A
+
+### Failure Modes
+The provided certs value will be validated before is it passed to the api-server flag
+
+Those conditions will cause Microshift to ignore certificates and log the error:
+1. certificates files exists in the disk and readable by microshift process.
+1. certificate should be parseable by [x509.ParseCertificate](https://pkg.go.dev/crypto/x509#ParseCertificate).
+1. certificate shouldn't override the internal certificates IPAddress/DNSNames in the SubjectAlternativeNames . see  [reserved](#reserved-names-values) names below.
+
+This condition will cause Microshift to accept the certificate with a warning in the logs.
+1. certificate is expired.
+
+
+### Reserved Names values
+| Address                   | Type     | Comment         |
+|:--------------------------|:---------|:----------------| 
+| localhost                 |DNS       |                 |
+| 127.0.0.1                 |IPAddress |                 |
```

Comment:
Do we need to add `::1` (ipv6)?

Reply:
do we  support IPv6 ?

Reply:
Hmm, not sure, I don't think so

Reply:
We have that on the roadmap, so let's try to avoid introducing any new places where we will have to do more work when it comes time to implementing it.